File: 0395-longest-substring-with-at-least-k-repeating-characters-至少有K个重复字符的最长子串/0395.py
class Solution:
    def longestSubstring(self, s: str, k: int) -> int:
        if len(s) < k: 
            return 0
        t = min(set(s), key=s.count)
        if s.count(t) >= k:
            return len(s)
        return max(self.longestSubstring(a, k) for a in s.split(t))

# 分治
# ref: https://leetcode-cn.com/problems/longest-substring-with-at-least-k-repeating-characters/solution/hua-dong-chuang-kou-fen-zhi-fa-by-powcai/

File: 0395-longest-substring-with-at-least-k-repeating-characters-至少有K个重复字符的最长子串/0395.cpp
// 分治
// ref: https://leetcode-cn.com/problems/longest-substring-with-at-least-k-repeating-characters/solution/hua-dong-chuang-kou-fen-zhi-fa-by-powcai/

class Solution {
public:
    int longestSubstring(string s, int k) {
        if (s.size() < k) return 0;
        // 查找字符串中出现次数最少的字母
        unordered_map<char, int> um;
        for (auto &a : s) {
            um[a]++;
        }
        int c, n = INT_MAX;
        for (auto &p : um) {
            if (p.second < n) {
                n = p.second;
                c = p.first;
            }
        }

        // 如果出现次数最少的字母个数大于k, 则当前子串都有效
        if (n >= k) return s.size();

        // 按出现次数个数最少的字母进行分割, 搜索(分而治之)
        string splited;
        int res = 0;
        for (int i = 0; i < s.size(); ++i) {
            if (s[i] == c) {
                res = max(longestSubstring(splited, k), res);
                splited = "";
            } else {
                splited.push_back(s[i]);
            }
        }
        res = max(longestSubstring(splited, k), res);
        return res;
    }
};